File: src/commands/show.ts
import { Logger } from "../logger.js";
import { ClangdClient } from "../clangd-client.js";
import { SymbolKind, Location } from "vscode-languageserver-protocol";
import * as fs from "fs";
import * as path from "path";
import { symbolKindToString, formatSymbolForDisplay } from "./utils.js";

/**
 * Helper to check if a file exists
 */
async function fileExists(filePath: string): Promise<boolean> {
  try {
    await fs.promises.access(filePath, fs.constants.F_OK);
    return true;
  } catch {
    return false;
  }
}

/**
 * Helper to check if a function/method has a body at the given location
 */
async function hasBody(client: ClangdClient, filePath: string, startLine: number): Promise<boolean> {
  const absolutePath = client.toAbsolutePath(filePath);
  const content = await fs.promises.readFile(absolutePath, 'utf-8');
  const lines = content.split('\n');
  
  // Check the next few lines for an opening brace
  for (let i = startLine; i < Math.min(startLine + 5, lines.length); i++) {
    if (lines[i].includes('{')) {
      return true;
    }
  }
  
  return false;
}

/**
 * Get contextual code around a symbol showing both declaration and definition.
 * This command intelligently handles C++ declaration/definition split.
 * 
 * For functions/methods, it shows BOTH:
 * - Declaration from header (with doc comments and signature)
 * - Definition from source (with implementation details)
 * 
 * Examples:
 * - "Scene::LoadResources" - shows declaration in .h and implementation in .cpp
 * - "ResourceManager" - shows class definition with context
 * - "CreateView" - shows function declaration and implementation
 * 
 * @param client - The ClangdClient instance
 * @param query - The symbol name to search for
 * @param logger - Logger for debug output
 * @returns Human-readable text with contextual code
 * @throws Error if no symbols match the query
 */
export async function getShowAsText(
  client: ClangdClient,
  query: string,
  logger: Logger
): Promise<string> {
  logger.info(`Getting context for: ${query}`);
  
  // Search for the symbol with a higher limit to ensure we find it
  const symbols = await client.searchSymbols(query, 50, logger);
  
  if (symbols.length === 0) {
    return `No symbols found matching "${query}"`;
  }
  
  // Use the best match - symbols are already sorted by relevance from clangd
  const symbol = symbols[0];
  
  const symbolKindName = symbolKindToString(symbol.kind);
  const fullName = formatSymbolForDisplay(symbol);
  
  // Get the symbol's location
  const symbolPath = client.pathFromFileUri(symbol.location.uri);
  const absoluteSymbolPath = client.toAbsolutePath(symbolPath);
  await client.ensureDocumentOpen(symbolPath);
  
  logger.debug(`Found ${symbolKindName} '${fullName}' at ${symbolPath}`);
  
  // For methods and functions, try to find both declaration and definition
  const locations: Array<{type: string, location: Location, path: string, isDefinition: boolean}> = [];
  
  if (symbol.kind === SymbolKind.Function || 
      symbol.kind === SymbolKind.Method || 
      symbol.kind === SymbolKind.Constructor) {
    
    // Determine if the symbol location is a definition or declaration
    const symbolHasBody = await hasBody(client, symbolPath, symbol.location.range.start.line);
    const symbolIsInSourceFile = symbolPath.match(/\.(cc|cpp|cxx|c\+\+)$/i) !== null;
    
    // In source files, it's almost always a definition
    // In header files, check if it has a body
    const symbolIsDefinition = symbolIsInSourceFile || symbolHasBody;
    
    locations.push({
      type: symbolIsDefinition ? 'definition' : 'declaration',
      location: symbol.location,
      path: symbolPath,
      isDefinition: symbolIsDefinition
    });
    
    // Get definition/declaration via textDocument/definition
    try {
      const definitions = await client.getDefinition(
        symbolPath,
        symbol.location.range.start.line,
        symbol.location.range.start.character,
        logger
      );
      
      logger.debug(`Found ${definitions.length} related location(s) via textDocument/definition`);
      
      // Add the other location(s) if different from current
      for (const def of definitions) {
        const defPath = client.pathFromFileUri(def.uri);
        
        // Skip if it's the same location
        if (defPath === symbolPath && 
            def.range.start.line === symbol.location.range.start.line) {
          continue;
        }
        
        // If we started from a definition, the other location is likely a declaration
        // If we started from a declaration, the other location is likely a definition
        let otherType: 'declaration' | 'definition';
        
        if (symbolIsDefinition) {
          // We have the definition, so the other location should be the declaration
          otherType = 'declaration';
        } else {
          // We have the declaration, check if the other location has a body
          const hasBodyAtDef = await hasBody(client, defPath, def.range.start.line);
          otherType = hasBodyAtDef ? 'definition' : 'declaration';
        }
        
        locations.push({
          type: otherType,
          location: def,
          path: defPath,
          isDefinition: otherType === 'definition'
        });
      }
    } catch (error) {
      logger.debug(`Failed to get related locations: ${error}`);
    }
    
    // Sort to show declaration first, then definition
    locations.sort((a, b) => {
      if (a.type === 'declaration' && b.type !== 'declaration') return -1;
      if (a.type !== 'declaration' && b.type === 'declaration') return 1;
      return 0;
    });
  } else {
    // For non-function symbols, just add the location
    locations.push({
      type: 'definition',
      location: symbol.location,
      path: symbolPath,
      isDefinition: true
    });
  }
  
  // Build the result
  let result = `Found ${symbolKindName} '${fullName}'`;
  if (symbols.length > 1) {
    result += ` (${symbols.length} matches total, showing most relevant)`;
  }
  result += '\n';
  
  // Get context for each location
  for (let i = 0; i < locations.length; i++) {
    const loc = locations[i];
    const absolutePath = client.toAbsolutePath(loc.path);
    
    // Read the file
    const content = await fs.promises.readFile(absolutePath, 'utf-8');
    const lines = content.split('\n');
    
    const startLine = loc.location.range.start.line;
    
    // Get folding ranges for this file to better understand code structure
    await client.ensureDocumentOpen(loc.path);
    const foldingRanges = await client.getFoldingRanges(loc.path, logger);
    
    let contextStart = startLine;
    let contextEnd = startLine;
    
    // Find preceding comments
    let commentStart = startLine;
    for (let j = startLine - 1; j >= 0 && j >= startLine - 50; j--) {
      const line = lines[j]?.trim() || '';
      if (line.startsWith('//') || line.startsWith('/*') || line.startsWith('*') || line === '*/') {
        commentStart = j;
      } else if (line !== '') {
        // Non-comment, non-empty line - stop
        break;
      }
    }
    
    // Handle functions/methods based on whether they are definitions
    if ((symbol.kind === SymbolKind.Function || 
         symbol.kind === SymbolKind.Method || 
         symbol.kind === SymbolKind.Constructor)) {
      
      if (loc.isDefinition) {
        // This is a definition, show the complete implementation
        contextStart = commentStart;
        
        // Find the folding range that represents this function body
        // Functions often have multiple folding ranges (signature, body, etc.)
        // We want the largest one that encompasses the entire function
        let functionRange = null;
        let bestRangeSize = 0;
        
        for (const range of foldingRanges) {
          // Check if this range starts at or near the function start
          // The opening brace might be on the same line or a few lines after
          if (range.startLine >= startLine - 1 && range.startLine <= startLine + 5) {
            const rangeSize = range.endLine - range.startLine;
            if (rangeSize > bestRangeSize) {
              functionRange = range;
              bestRangeSize = rangeSize;
            }
          }
        }
        
        if (functionRange) {
          // Folding ranges sometimes don't include the closing brace
          // Add 1 to ensure we include it
          contextEnd = Math.min(functionRange.endLine + 1, lines.length - 1);
        } else {
          // Fallback: manual brace counting
          let braceCount = 0;
          let foundOpeningBrace = false;
          
          for (let j = startLine; j < lines.length && j < startLine + 100; j++) {
            const line = lines[j];
            for (const char of line) {
              if (char === '{') {
                braceCount++;
                foundOpeningBrace = true;
              } else if (char === '}') {
                braceCount--;
              }
            }
            
            contextEnd = j;
            
            if (foundOpeningBrace && braceCount === 0) {
              break;
            }
          }
        }
      } else {
        // This is a declaration (no body), show declaration + a bit of context
        contextStart = commentStart;
        contextEnd = loc.location.range.end.line;
        
        // Include a couple lines after to show the next method/member for context
        contextEnd = Math.min(contextEnd + 2, lines.length - 1);
      }
    }
    
    // For other types (classes, structs, etc), use folding ranges if available
    else if (symbol.kind === SymbolKind.Class || 
             symbol.kind === SymbolKind.Struct || 
             symbol.kind === SymbolKind.Enum) {
      
      // Use comments we already found
      contextStart = commentStart;
      
      // Use folding range for the body
      let classRange = null;
      for (const range of foldingRanges) {
        if (range.startLine >= startLine && range.startLine <= startLine + 2) {
          classRange = range;
          break;
        }
      }
      
      if (classRange) {
        // For classes, show up to 30 lines to give a good overview
        contextEnd = Math.min(classRange.endLine, startLine + 30);
      } else {
        // Fallback: show 20 lines for class/struct/enum definitions
        contextEnd = Math.min(startLine + 20, lines.length - 1);
      }
    }
    // For other symbol types (variables, typedefs, etc)
    else {
      contextStart = commentStart;
      contextEnd = loc.location.range.end.line;
      // Show a few more lines for context
      contextEnd = Math.min(contextEnd + 5, lines.length - 1);
    }
    
    // Extract the context lines
    const extractedLines = lines.slice(contextStart, contextEnd + 1);
    
    // Format the section header
    result += '\n';
    const formattedLoc = client.formatLocation(loc.path, startLine, 0);
    
    // Always show the type for functions/methods/constructors
    if (symbol.kind === SymbolKind.Function || 
        symbol.kind === SymbolKind.Method || 
        symbol.kind === SymbolKind.Constructor) {
      if (loc.type === 'declaration') {
        result += `From ${formattedLoc} (declaration)\n`;
      } else if (loc.type === 'definition') {
        result += `From ${formattedLoc} (definition)\n`;
      } else {
        result += `From ${formattedLoc}\n`;
      }
    } else {
      result += `From ${formattedLoc}\n`;
    }
    
    // Add the code block
    result += '```cpp\n';
    result += extractedLines.join('\n');
    result += '\n```';
    
    if (i < locations.length - 1) {
      result += '\n';
    }
  }
  
  return result;
}

File: go/internal/commands/show.go
package commands

import (
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"

	"clangd-query/internal/logger"
	"clangd-query/internal/lsp"
)

// Show displays both declaration and definition of a symbol with contextual code
// This command intelligently handles C++ declaration/definition split.
func Show(client *lsp.ClangdClient, query string, log logger.Logger) (string, error) {
	log.Info("Getting context for: %s", query)

	// Search for the symbol with a higher limit to ensure we find it
	symbols, err := client.WorkspaceSymbol(query)
	if err != nil {
		return "", err
	}

	if len(symbols) == 0 {
		return fmt.Sprintf(`No symbols found matching "%s"`, query), nil
	}

	// Use the best match - symbols are already sorted by relevance from clangd
	symbol := symbols[0]

	symbolKindName := SymbolKindToString(symbol.Kind)
	fullName := formatSymbolForDisplay(symbol)

	// Get the symbol's location
	symbolPath := client.PathFromFileURI(symbol.Location.URI)
	
	log.Debug("Found %s '%s' at %s", symbolKindName, fullName, symbolPath)

	// For methods and functions, try to find both declaration and definition
	type locationInfo struct {
		locType      string
		location     lsp.Location
		path         string
		isDefinition bool
	}

	var locations []locationInfo

	if symbol.Kind == lsp.SymbolKindFunction || 
	   symbol.Kind == lsp.SymbolKindMethod || 
	   symbol.Kind == lsp.SymbolKindConstructor {

		// Determine if the symbol location is a definition or declaration
		symbolHasBody, _ := hasBody(symbolPath, symbol.Location.Range.Start.Line)
		symbolIsInSourceFile := regexp.MustCompile(`\.(cc|cpp|cxx|c\+\+)$`).MatchString(strings.ToLower(symbolPath))

		// In source files, it's almost always a definition
		// In header files, check if it has a body
		symbolIsDefinition := symbolIsInSourceFile || symbolHasBody

		locType := "declaration"
		if symbolIsDefinition {
			locType = "definition"
		}

		locations = append(locations, locationInfo{
			locType:      locType,
			location:     symbol.Location,
			path:         symbolPath,
			isDefinition: symbolIsDefinition,
		})

		// Get definition/declaration via textDocument/definition
		definitions, err := client.GetDefinition(symbol.Location.URI, symbol.Location.Range.Start)
		if err == nil {
			log.Debug("Found %d related location(s) via textDocument/definition", len(definitions))

			// Add the other location(s) if different from current
			for _, def := range definitions {
				defPath := client.PathFromFileURI(def.URI)

				// Skip if it's the same location
				if defPath == symbolPath && 
				   def.Range.Start.Line == symbol.Location.Range.Start.Line {
					continue
				}

				// If we started from a definition, the other location is likely a declaration
				// If we started from a declaration, the other location is likely a definition
				var otherType string

				if symbolIsDefinition {
					// We have the definition, so the other location should be the declaration
					otherType = "declaration"
				} else {
					// We have the declaration, check if the other location has a body
					hasBodyAtDef, _ := hasBody(defPath, def.Range.Start.Line)
					if hasBodyAtDef {
						otherType = "definition"
					} else {
						otherType = "declaration"
					}
				}

				locations = append(locations, locationInfo{
					locType:      otherType,
					location:     def,
					path:         defPath,
					isDefinition: otherType == "definition",
				})
			}
		} else {
			log.Debug("Failed to get related locations: %v", err)
		}

		// Sort to show declaration first, then definition
		sort.Slice(locations, func(i, j int) bool {
			if locations[i].locType == "declaration" && locations[j].locType != "declaration" {
				return true
			}
			if locations[i].locType != "declaration" && locations[j].locType == "declaration" {
				return false
			}
			return false
		})
	} else {
		// For non-function symbols, just add the location
		locations = append(locations, locationInfo{
			locType:      "definition",
			location:     symbol.Location,
			path:         symbolPath,
			isDefinition: true,
		})
	}

	// Build the result
	result := fmt.Sprintf("Found %s '%s'", symbolKindName, fullName)
	if len(symbols) > 1 {
		result += fmt.Sprintf(" (%d matches total, showing most relevant)", len(symbols))
	}
	result += "\n"

	// Get context for each location
	for i, loc := range locations {
		// Read the file
		content, err := os.ReadFile(loc.path)
		if err != nil {
			log.Debug("Failed to read file %s: %v", loc.path, err)
			continue
		}
		lines := strings.Split(string(content), "\n")

		startLine := loc.location.Range.Start.Line

		// Get folding ranges for this file to better understand code structure
		foldingRanges, err := client.GetFoldingRanges(loc.location.URI)
		if err != nil {
			log.Debug("Failed to get folding ranges: %v", err)
			foldingRanges = []lsp.FoldingRange{}
		}

		contextStart := startLine
		contextEnd := startLine

		// Find preceding comments
		commentStart := startLine
		for j := startLine - 1; j >= 0 && j >= startLine-50; j-- {
			if j >= len(lines) {
				continue
			}
			line := strings.TrimSpace(lines[j])
			if strings.HasPrefix(line, "//") || strings.HasPrefix(line, "/*") || 
			   strings.HasPrefix(line, "*") || line == "*/" {
				commentStart = j
			} else if line != "" {
				// Non-comment, non-empty line - stop
				break
			}
		}

		// Handle functions/methods based on whether they are definitions
		if (symbol.Kind == lsp.SymbolKindFunction || 
		    symbol.Kind == lsp.SymbolKindMethod || 
		    symbol.Kind == lsp.SymbolKindConstructor) {

			if loc.isDefinition {
				// This is a definition, show the complete implementation
				contextStart = commentStart

				// Find the folding range that represents this function body
				var functionRange *lsp.FoldingRange
				bestRangeSize := 0

				for _, foldRange := range foldingRanges {
					// Check if this range starts at or near the function start
					// The opening brace might be on the same line or a few lines after
					if foldRange.StartLine >= startLine-1 && foldRange.StartLine <= startLine+5 {
						rangeSize := foldRange.EndLine - foldRange.StartLine
						if rangeSize > bestRangeSize {
							functionRange = &foldRange
							bestRangeSize = rangeSize
						}
					}
				}

				if functionRange != nil {
					// Folding ranges sometimes don't include the closing brace
					// Add 1 to ensure we include it
					contextEnd = min(functionRange.EndLine+1, len(lines)-1)
				} else {
					// If no folding range found, show a reasonable amount
					contextEnd = min(startLine+50, len(lines)-1)
				}
			} else {
				// This is a declaration (no body), show just the declaration
				contextStart = commentStart
				contextEnd = loc.location.Range.End.Line
			}
		} else if symbol.Kind == lsp.SymbolKindClass || 
		          symbol.Kind == lsp.SymbolKindStruct || 
		          symbol.Kind == lsp.SymbolKindEnum {
			// Use comments we already found
			contextStart = commentStart

			// Use folding range for the body
			var classRange *lsp.FoldingRange
			for _, foldRange := range foldingRanges {
				if foldRange.StartLine >= startLine && foldRange.StartLine <= startLine+2 {
					classRange = &foldRange
					break
				}
			}

			if classRange != nil {
				// For classes, show the complete implementation
				contextEnd = classRange.EndLine
			} else {
				// If no folding range found, show a reasonable amount
				contextEnd = min(startLine+100, len(lines)-1)
			}
		} else {
			// For other symbol types (variables, typedefs, etc)
			contextStart = commentStart
			contextEnd = loc.location.Range.End.Line
		}

		// Ensure bounds are valid
		contextStart = max(0, contextStart)
		contextEnd = min(contextEnd, len(lines)-1)

		// Extract the context lines
		var extractedLines []string
		if contextStart <= contextEnd {
			extractedLines = lines[contextStart:contextEnd+1]
		}

		// Format the section header
		result += "\n"
		// Use formatLocation with full location info including column
		formattedLoc := formatLocation(client, loc.location)

		// Always show the type for functions/methods/constructors
		if symbol.Kind == lsp.SymbolKindFunction || 
		   symbol.Kind == lsp.SymbolKindMethod || 
		   symbol.Kind == lsp.SymbolKindConstructor {
			if loc.locType == "declaration" {
				result += fmt.Sprintf("From %s (declaration)\n", formattedLoc)
			} else if loc.locType == "definition" {
				result += fmt.Sprintf("From %s (definition)\n", formattedLoc)
			} else {
				result += fmt.Sprintf("From %s\n", formattedLoc)
			}
		} else {
			result += fmt.Sprintf("From %s\n", formattedLoc)
		}

		// Add the code block
		result += "```cpp\n"
		result += strings.Join(extractedLines, "\n")
		result += "\n```"

		if i < len(locations)-1 {
			result += "\n"
		}
	}

	return result, nil
}

// hasBody checks if a function/method has a body at the given location
func hasBody(filePath string, startLine int) (bool, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return false, err
	}
	
	lines := strings.Split(string(content), "\n")
	
	// Check the next few lines for an opening brace
	for i := startLine; i < min(startLine+5, len(lines)); i++ {
		if i < len(lines) && strings.Contains(lines[i], "{") {
			return true, nil
		}
	}
	
	return false, nil
}


// min returns the minimum of two integers
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// max returns the maximum of two integers  
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

